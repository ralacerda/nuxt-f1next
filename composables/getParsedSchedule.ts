import { formatTimeAgo } from "@vueuse/core";

export default function (
  eventSchedule: { name: string; datetime: string }[],
  round: number,
  nextRound: number
) {
  const date = new Date(Date.now());
  const result = eventSchedule
    .map((event) => ({
      ...event,
      datetime: new Date(event.datetime),
    }))
    .sort((a, b) => a.datetime.getTime() - b.datetime.getTime())
    .map((event) => ({
      ...event,
      relativeTime: formatTimeAgo(event.datetime, { rounding: "floor" }, date),
      state: event.datetime > date ? "future" : "past",
    }));
  if (round == nextRound) {
    const next = result.findIndex((event) => event.state == "future");
    result[next].state = "next";
  }
  return result;
}
