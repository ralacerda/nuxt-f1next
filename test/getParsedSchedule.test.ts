import { describe, test, expect, vi } from "vitest";
import getParsedSchedule from "composables/getParsedSchedule";

describe("Parse the schedule", () => {
  vi.setSystemTime("2023-03-04T14:26:21.000Z");
  test("Get the round if the race is in the same day", () => {
    const result = getParsedSchedule(schedule.schedule, 1, 1);
    expect(result).toMatchSnapshot();
  });
});

const schedule = {
  country: "Bahrain",
  schedule: [
    { name: "Race", datetime: "2023-03-05T15:00:00Z" },
    { name: "First Practice", datetime: "2023-03-03T11:30:00Z" },
    { name: "Second Practice", datetime: "2023-03-03T15:00:00Z" },
    { name: "Third Practice", datetime: "2023-03-04T11:30:00Z" },
    { name: "Qualifying", datetime: "2023-03-04T15:00:00Z" },
  ],
  round: "1",
  date: "2023-03-05",
  raceName: "Bahrain Grand Prix",
};
