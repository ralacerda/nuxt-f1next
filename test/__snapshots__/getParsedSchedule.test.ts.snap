// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parse the schedule > Do not indicate the next event when the current round is the next round 1`] = `
[
  {
    "datetime": 2023-03-03T11:30:00.000Z,
    "name": "First Practice",
    "relativeTime": "yesterday",
    "state": "past",
  },
  {
    "datetime": 2023-03-03T15:00:00.000Z,
    "name": "Second Practice",
    "relativeTime": "23 hours ago",
    "state": "past",
  },
  {
    "datetime": 2023-03-04T11:30:00.000Z,
    "name": "Third Practice",
    "relativeTime": "2 hours ago",
    "state": "past",
  },
  {
    "datetime": 2023-03-04T15:00:00.000Z,
    "name": "Qualifying",
    "relativeTime": "in 33 minutes",
    "state": "future",
  },
  {
    "datetime": 2023-03-05T15:00:00.000Z,
    "name": "Race",
    "relativeTime": "tomorrow",
    "state": "future",
  },
]
`;

exports[`Parse the schedule > Indicate the next event when the current round is the next round 1`] = `
[
  {
    "datetime": 2023-03-03T11:30:00.000Z,
    "name": "First Practice",
    "relativeTime": "yesterday",
    "state": "past",
  },
  {
    "datetime": 2023-03-03T15:00:00.000Z,
    "name": "Second Practice",
    "relativeTime": "23 hours ago",
    "state": "past",
  },
  {
    "datetime": 2023-03-04T11:30:00.000Z,
    "name": "Third Practice",
    "relativeTime": "2 hours ago",
    "state": "past",
  },
  {
    "datetime": 2023-03-04T15:00:00.000Z,
    "name": "Qualifying",
    "relativeTime": "in 33 minutes",
    "state": "next",
  },
  {
    "datetime": 2023-03-05T15:00:00.000Z,
    "name": "Race",
    "relativeTime": "tomorrow",
    "state": "future",
  },
]
`;
